### PR TITLE
DIG-1131: Refactoring federation 

### PR DESCRIPTION
**Pull request by daisieh**
So much stuff has changed. The documentation should be pretty much all inside the federation.yaml file, which you can view as a swagger-ui thingy at http://candig.docker.internal:4232/federation/v1/. I deleted the Sphinx doc system because it was just too much baggage for a package with, like, four endpoints.

The endpoints are now:
- /federation/v1/servers (and /servers/id): these list, add, delete, get the registered peer servers
- /federation/v1/services (and /services/id): these list, add, delete, get the registered services
- /federation/v1/service-info: this gets some basic info about the federation service itself, including the host server
- /federation/v1/fanout: this is the actual endpoint that makes the federated requests to other services. The request body has been simplified; check the swagger API for more information.

I updated the README with current information.

This module can be tested on its own:
```
make build-federation
make compose-federation
docker exec candigv2_federation_1 pytest
```

### CODE REVIEW

**Comment by daisieh** [comment]
In https://github.com/CanDIG/CanDIGv2/pull/231#issuecomment-1552138508, Son had problems building federation. It looks like the error originates because of the new candigv2-authx requirement of cryptography. I'm going to add another package to the Dockerfile which should help.

**Comment by SonQBChau** [comment]
I make clean-all and build-all, still having another error:
```
make[2]: [compose-federation] Error 1 (ignored)
source /Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/setup_hosts.sh; \
        docker compose -f /Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/lib/candigv2/docker-compose.yml -f /Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/lib/federation/docker-compose.yml --compatibility up -d
IP Address is already set to: 192.168.4.69
WARN[0000] Found orphan containers ([candigv2_vault_1 candigv2_vault-runner_1 candigv2_opa-runner_1 candigv2_opa_1 candigv2_tyk_1 candigv2_tyk-redis_1 candigv2_keycloak_1 candigv2_candig-data-portal_1 candigv2_katsu_1 candigv2_metadata-db_1 candigv2_htsget_1 candigv2_minio_1]) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up. 
[+] Running 1/0
 ✔ Container candigv2_federation_1  Running                                                                                                                    0.0s 
source /Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/lib/federation/federation_setup.sh
>> waiting for federation to start
candigv2_federation_1
Traceback (most recent call last):
  File "/Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/lib/federation/initialize.py", line 65, in <module>
    main()
  File "/Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/lib/federation/initialize.py", line 30, in main
    token = get_access_token(username=os.getenv("CANDIG_SITE_ADMIN_USER"), password=os.getenv("CANDIG_SITE_ADMIN_PASSWORD"))
  File "/Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/bin/miniconda3/envs/candig/lib/python3.10/site-packages/authx/auth.py", line 65, in get_access_token
    raise Exception(f"Check for environment variables: {response.text}")
Exception: Check for environment variables: {"error":"unauthorized_client","error_description":"Invalid client secret"}
make[2]: [compose-federation] Error 1 (ignored)
echo "    finished compose-federation" >> /Users/sonchau/Documents/GitHub/CANDIG/CanDIGv2/tmp/progress.txt
```

File: README.md [hunk 1]
```diff
@@ -19,57 +19,30 @@ The federation_service can be installed in a py3.7+ virtual environment:
...
-
-In addition, specify the location for your peer servers with the array `["research centre", "province name", "province code"]`. For compatibility with CanDIG's [data portal](https://github.com/CanDIG/candig-data-portal), use the following province codes:
-
-`'ca-ab', 'ca-bc', 'ca-mb', 'ca-nb', 'ca-nl', 'ca-nt', 'ca-ns', 'ca-nu', 'ca-on', 'ca-pe', 'ca-qc', 'ca-sk', 'ca-yt'`
-
-
-For example, if your host federation service is running at `http://0.0.0.0:8890` in British Columbia, and your first 
-peer server federation service is running at `http://0.0.0.0:8891` in Ontario, your `servers.json` would look like this:
+On initialization of the docker container, the server listed in .env as FEDERATION_SELF_SERVER will be registered. This is your own server. If you want to register other peer servers, use the /federation/v1/servers POST endpoint, described in federation.yaml. You'll need to have a valid JWT from the peer server's identity issuer, as well as the URL of the issuer. This call must be authorized with a bearer token from a site administrator.
```
Comment (kcranston) [inline README.md]: Can you be explicit about the two pieces of information that we need from the JWT in order to set up a new server? Also, be specific that the admin bearer token is from the local instance, not the peer being added.

**Comment by daisieh** [comment]
@SonQBChau this one is related to a recent commit in https://github.com/CanDIG/CanDIGv2/pull/231#issuecomment-1553379559

File: README.md (hunk 1 above)
Comment (daisieh) [reply README.md]: https://github.com/CanDIG/federation_service/pull/36/commits/8be405e886904bb37ad350a56bde0996849d6c6e

Is that more clear?

**Review by SonQBChau — APPROVED** [review]
worked for me

**Pull request merged**